### ISSUE TITLE
asteval and add_param_hint() error

### ISSUE THREAD

**Issue by tritemio** [lmfit/lmfit-py#271]
I get an error when I add a parameter to a `CompositeModel` through `.add_param_hint()`.

This was tested the latest lmfit branch: https://github.com/lmfit/lmfit-py/pull/270

Python 3.4, Windows 7 x64.

Here the full example:

```
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import lmfit
print(lmfit.__version__)

data = pd.DataFrame({'counts': {
    -0.1875: 0, -0.1625: 0, -0.1375: 4, -0.1125: 2, -0.0875: 1,
    -0.0625: 8, -0.0375: 21, -0.0125: 21,  0.0125: 23,  0.0375: 35,
     0.0625: 50,  0.0875: 54,  0.1125: 46,  0.1375: 70,  0.1625: 77,
     0.1875: 87,  0.2125: 98,  0.2375: 113,  0.2625: 148,  0.2875: 136,
     0.3125: 185,  0.3375: 195,  0.3625: 194,  0.3875: 168,  0.4125: 170,
     0.4375: 139,  0.4625: 155,  0.4875: 115,  0.5125: 132,  0.5375: 109,
     0.5625: 102,  0.5875: 85,  0.6125: 69,  0.6375: 81,  0.6625: 82,
     0.6875: 80,  0.7125: 71,  0.7375: 64,  0.7625: 79,  0.7875: 88,
     0.8125: 111,  0.8375: 97,  0.8625: 97,  0.8875: 73,  0.9125: 72,
     0.9375: 62,  0.9625: 41,  0.9875: 30,  1.0125: 13,  1.0375: 3,
     1.0625: 9,  1.0875: 7,  1.1125: 0,  1.1375: 0,  1.1625: 0, }})
#plt.plot(data)

def gauss(x, sigma, mu, A):
    return A*np.exp(-(x-mu)**2/(2*sigma**2))

p1_center = 0.2
p1_sigma = 0.1
p2_center = 0.8
p2_sigma = 0.1

peak1 = lmfit.model.Model(gauss, prefix='p1_')
peak2 = lmfit.model.Model(gauss, prefix='p2_')
model = peak1 + peak2

model.set_param_hint('p1_center', value=p1_center, min=-1, max=2)
model.set_param_hint('p2_center', value=p2_center, min=-1, max=2)
model.set_param_hint('p1_sigma1', value=p1_sigma, min=0.01, max=0.2)
model.set_param_hint('p1_sigma2', value=p1_sigma, min=0.01, max=0.2)
model.set_param_hint('p2_sigma1', value=p2_sigma, min=0.01, max=0.2)
model.set_param_hint('p2_sigma2', value=p2_sigma, min=0.01, max=0.2)
model.set_param_hint('p1_amplitude', value=1, min=0.01)
model.set_param_hint('p2_amplitude', value=1, min=0.01)

# This constrains the distance between peaks to be > 0
model.set_param_hint('pos_delta', value=0.3, min=0)
model.set_param_hint('p2_center', min=-1, expr='p1_center + pos_delta')

model.fit(data['counts'], x=data.index)

```

The error is the following:

```
NameError
   <_ast.Module object at 0x0000000007070F98>
                ^^^
name 'pos_delta' is not defined
Traceback (most recent call last):
  File "lmfit_0.9.2rc.py", line 47, in <module>
    model.fit(data['counts'], x=data.index)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\model.py", line 440, in fit
    params = self.make_params(verbose=verbose)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\model.py", line 304, in make_params
    params.add_many(*_params)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 231, in add_many
    self.__setitem__(para.name, para)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 85, in __setitem__
    self._asteval.symtable[key] = par.value
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 556, in value
    return self._getval()
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 532, in _getval
    check_ast_errors(self._expr_eval)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 22, in check_ast_errors
    expr_eval.raise_exception(None)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\asteval.py", line 167, in raise_exception
    raise exc(self.error_msg)
NameError: name 'pos_delta' is not defined in expr='<_ast.Module object at 0x0000000007070F98>'
```

**EDIT**: typos

**Comment by newville** [comment]
@tritemio Ok, thanks!  Will fix.  And that should be in a test!

**Comment by tritemio** [comment]
@newville I adapted the example into a test which fails with the same error.

**Comment by tritemio** [comment]
Still failing with same error using https://github.com/lmfit/lmfit-py/commit/9304f42c2d1bf7cff3e3a471ae7263e512f96c19:

```
0.9.1-49-g9304f42
 - Adding parameter for hint "p1_center"
NameError
   <_ast.Module object at 0x0000000007121D30>
                ^^^
name 'pos_delta' is not defined
Traceback (most recent call last):
  File "lmfit_0.9.2rc.py", line 53, in <module>
    model.fit(y, x=x)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\model.py", line 434, in fit
    params = self.make_params(verbose=verbose)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\model.py", line 269, in make_params
    params.add(par)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 197, in add
    self.__setitem__(name.name, name)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 85, in __setitem__
    self._asteval.symtable[key] = par.value
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 556, in value
    return self._getval()
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 532, in _getval
    check_ast_errors(self._expr_eval)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\parameter.py", line 22, in check_ast_errors
    expr_eval.raise_exception(None)
  File "C:\Users\laser2002j\Anaconda\envs\py3\lib\site-packages\lmfit\asteval.py", line 167, in raise_exception
    raise exc(self.error_msg)
NameError: name 'pos_delta' is not defined in expr='<_ast.Module object at 0x0000000007121D30>'
```

**Comment by newville** [comment]
@tritemio Ah, this is now because you have multiple hints for the same parameter. 

```python
model.set_param_hint('p2_center', value=p2_center, min=-1, max=2)
....
model.set_param_hint('pos_delta', value=0.3, min=0)
model.set_param_hint('p2_center', min=-1, expr='p1_center + pos_delta')
```

This has the unfortunate effect of `Model.make_params` making a Parameter for `p2_center` with the expression *before* `pos_delta` has been defined.  You won't get such an error if you remove the first hint for `p2_center`.    You will find other problems, though ;).

In general I think it's OK to give an error when an expression is defined that cannot be evaluated immediately, but it is slightly weird for a model's parameter hints to fail in this way.   My inclination would be to make `set_parameter_hint()` replace and *not* update the hint.    But that could break existing scripts that had relied on 

```python 
model.set_param_hint('var', min=0)
model.set_param_hint('var', max=100)
```
to be the same as
```python 
model.set_param_hint('var', min=0, max=100)
```

but I don't think we ever guaranteed that those were, in fact, equivalent.   

I definitely view this as "hints gone wild".  You have a lot of parameter hints here, and most of them are clearly not going to be used, including the ones causing this trouble.   And, of course, this fit is not going to work, as the names for all the hints are wrong, and you're not setting any hints for or giving initial values for the actual parameters.   In fact, your script sort of shows to me a good example of why using hints to automatically create Parameters is dangerous, needs to be done with great care, and should not replace actually creating Parameters.  It's not the error in the expression that demonstrates this, but the fact that you don't even create the correct parameters.  Whereas, if you had done
```python
peak1 = lmfit.model.Model(gauss, prefix='p1_')
peak2 = lmfit.model.Model(gauss, prefix='p2_')
model = peak1 + peak2
params = model.make_params()

params['pos_delta'] = lmfit.Parameter(value=0.5, min=0.2, max=1)
params['p1_mu'] = lmfit.Parameter(value=0.2, min=-1, max=2)
params['p2_mu'] = lmfit.Parameter(expr='p1_mu + pos_delta')
params['p1_sigma'] = lmfit.Parameter(value=0.2, min=0.01)
params['p2_sigma'] = lmfit.Parameter(value=0.2, min=0.01)
params['p1_A'] = lmfit.Parameter(value=100, min=0.01)
params['p2_A'] = lmfit.Parameter(value=100, min=0.01)

x = np.array(data.index)
out = model.fit(data['counts'], params, x=x)

print out.fit_report()
```

the fit would work reasonably well.

OK, I guess that the error message here is not ideal, but the answer might be "don't do that".   The parameter hints for a model are a real pain -- maybe we should work toward a new way to set such "hints".  So, the solutions I see are:
  1. make `model.set_param_hint()` on an existing parameter name overwrite the previous hint.
  2. just give up and say that "of course, you *can* get errors if your constraint expressions cannot be evaluated" and "using lots of parameter hints instead of making parameters is a good way to find problems you did not intend".

I think either of these options is fine.  If we could break more scripts like this one, that might be a positive outcome.

Any ideas on what the right behavior should be?

**Comment by tritemio** [comment]
@newville yeah sorry the example has may other mistakes, it was a reduced from a larger script in which the peak is more complex and I messed up.

Anyway, I think solution 1 is fine.

Also, let me explain why I'm using `set_param_hints`. I create the model in a function, set the hints and return the model (factory function. The model itself needs to have specific constraints on several parameters. For example there is a range of physically acceptable peak widths values. Additionally I want to assure that the `p1_` peak always fits the peak at smaller `x`, `p2_` the next peak and so on. You can see a real working example [here](https://github.com/tritemio/FRETBursts/blob/0.4.7/fretbursts/mfit.py#L228).

Using your example the factory function should return a model and a set of parameters with constraints which, for an end-user, is kind of strange (why a model cannot contain its own parameters?).

**Comment by newville** [comment]
@tritemio Yeah, I think it's fine to use parameter hints in a "factory function" setting, but there you'd be testing that and careful to avoid mistakes, so giving errors because of expressions can't yet be evaluated might be OK.  

As an aside, I'm also willing to say that maybe we should allow "out-of-order" definitions and constraint expressions until they are actually needed.  The question then is "how do we know when they are actually needed?".  There was an earlier conversation in which the preferred answer was "as soon as possible".  

Anyway, we'll go with option 1 (a parameter hint replaces a previous definition, or at least goes to the end of the list of hints).

**Comment by newville** [comment]
addressed with 208871db96689c911e00b9d59dace2de68df3220
